File: task3-unknown-language/noteapp/internal/repository/note_repository.go
package repository

import (
	"database/sql"
	"fmt"
	"noteapp/internal/model"
	"time"
)

// NoteRepository интерфейс для работы с заметками в базе данных
type NoteRepository interface {
	GetAll() ([]*model.Note, error)
	GetByID(id int) (*model.Note, error)
	Create(text string) (*model.Note, error)
	Update(id int, text string) (*model.Note, error)
	Delete(id int) error
}

// noteRepository реализация репозитория для заметок
type noteRepository struct {
	db *sql.DB
}

// NewNoteRepository создает новый экземпляр репозитория заметок
func NewNoteRepository(db *sql.DB) NoteRepository {
	return &noteRepository{db: db}
}

// GetAll возвращает все заметки из базы данных
func (r *noteRepository) GetAll() ([]*model.Note, error) {
	query := "SELECT id, text, created_at FROM notes ORDER BY created_at DESC"
	rows, err := r.db.Query(query)
	if err != nil {
		return nil, fmt.Errorf("failed to query notes: %w", err)
	}
	defer rows.Close()

	var notes []*model.Note
	for rows.Next() {
		note := &model.Note{}
		err := rows.Scan(&note.ID, &note.Text, &note.CreatedAt)
		if err != nil {
			return nil, fmt.Errorf("failed to scan note: %w", err)
		}
		notes = append(notes, note)
	}

	if err = rows.Err(); err != nil {
		return nil, fmt.Errorf("rows iteration error: %w", err)
	}

	return notes, nil
}

// GetByID возвращает заметку по ID
func (r *noteRepository) GetByID(id int) (*model.Note, error) {
	query := "SELECT id, text, created_at FROM notes WHERE id = ?"
	row := r.db.QueryRow(query, id)

	note := &model.Note{}
	err := row.Scan(&note.ID, &note.Text, &note.CreatedAt)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, model.ErrNoteNotFound
		}
		return nil, fmt.Errorf("failed to scan note: %w", err)
	}

	return note, nil
}

// Create создает новую заметку
func (r *noteRepository) Create(text string) (*model.Note, error) {
	query := "INSERT INTO notes (text, created_at) VALUES (?, ?)"
	createdAt := time.Now()
	
	result, err := r.db.Exec(query, text, createdAt)
	if err != nil {
		return nil, fmt.Errorf("failed to create note: %w", err)
	}

	id, err := result.LastInsertId()
	if err != nil {
		return nil, fmt.Errorf("failed to get last insert id: %w", err)
	}

	return &model.Note{
		ID:        int(id),
		Text:      text,
		CreatedAt: createdAt,
	}, nil
}

// Update обновляет существующую заметку
func (r *noteRepository) Update(id int, text string) (*model.Note, error) {
	// Сначала проверим, существует ли заметка
	_, err := r.GetByID(id)
	if err != nil {
		return nil, err
	}

	query := "UPDATE notes SET text = ? WHERE id = ?"
	_, err = r.db.Exec(query, text, id)
	if err != nil {
		return nil, fmt.Errorf("failed to update note: %w", err)
	}

	// Возвращаем обновленную заметку
	return r.GetByID(id)
}

// Delete удаляет заметку по ID
func (r *noteRepository) Delete(id int) error {
	// Сначала проверим, существует ли заметка
	_, err := r.GetByID(id)
	if err != nil {
		return err
	}

	query := "DELETE FROM notes WHERE id = ?"
	_, err = r.db.Exec(query, id)
	if err != nil {
		return fmt.Errorf("failed to delete note: %w", err)
	}

	return nil
}

File: task3-unknown-language/Python-notes-app/repository/note_repository.py
"""
Репозиторий для операций с заметками в базе данных.
Содержит методы для CRUD операций с таблицей notes.
"""

import sqlite3
from datetime import datetime
from typing import List, Optional
from db.database import execute_query
from model.note import Note


class NoteRepository:
    """Репозиторий для работы с заметками в базе данных."""
    
    def get_all(self) -> List[Note]:
        """
        Получает все заметки из базы данных.
        
        Returns:
            Список всех заметок
        """
        try:
            query = "SELECT id, text, created_at FROM notes ORDER BY created_at DESC"
            rows = execute_query(query)
            
            notes = []
            for row in rows:
                note = Note(
                    id=row['id'],
                    text=row['text'],
                    created_at=datetime.fromisoformat(row['created_at']) if row['created_at'] else None
                )
                notes.append(note)
            
            return notes
            
        except sqlite3.Error as e:
            raise Exception(f"Ошибка при получении заметок: {e}")
    
    def get_by_id(self, note_id: int) -> Optional[Note]:
        """
        Получает заметку по ID.
        
        Args:
            note_id: ID заметки
            
        Returns:
            Заметка или None, если не найдена
        """
        try:
            query = "SELECT id, text, created_at FROM notes WHERE id = ?"
            rows = execute_query(query, (note_id,))
            
            if not rows:
                return None
            
            row = rows[0]
            return Note(
                id=row['id'],
                text=row['text'],
                created_at=datetime.fromisoformat(row['created_at']) if row['created_at'] else None
            )
            
        except sqlite3.Error as e:
            raise Exception(f"Ошибка при получении заметки с ID {note_id}: {e}")
    
    def create(self, note: Note) -> Note:
        """
        Создает новую заметку в базе данных.
        
        Args:
            note: Объект заметки для создания
            
        Returns:
            Созданная заметка с присвоенным ID
        """
        try:
            query = "INSERT INTO notes (text, created_at) VALUES (?, ?)"
            created_at = datetime.now()
            note_id = execute_query(query, (note.text, created_at.isoformat()))
            
            return Note(
                id=note_id,
                text=note.text,
                created_at=created_at
            )
            
        except sqlite3.Error as e:
            raise Exception(f"Ошибка при создании заметки: {e}")
    
    def update(self, note_id: int, note: Note) -> Optional[Note]:
        """
        Обновляет существующую заметку.
        
        Args:
            note_id: ID заметки для обновления
            note: Новые данные заметки
            
        Returns:
            Обновленная заметка или None, если заметка не найдена
        """
        try:
            # Проверяем, существует ли заметка
            existing_note = self.get_by_id(note_id)
            if not existing_note:
                return None
            
            query = "UPDATE notes SET text = ? WHERE id = ?"
            execute_query(query, (note.text, note_id))
            
            # Возвращаем обновленную заметку
            return self.get_by_id(note_id)
            
        except sqlite3.Error as e:
            raise Exception(f"Ошибка при обновлении заметки с ID {note_id}: {e}")
    
    def delete(self, note_id: int) -> bool:
        """
        Удаляет заметку по ID.
        
        Args:
            note_id: ID заметки для удаления
            
        Returns:
            True, если заметка была удалена, False если не найдена
        """
        try:
            # Проверяем, существует ли заметка
            existing_note = self.get_by_id(note_id)
            if not existing_note:
                return False
            
            query = "DELETE FROM notes WHERE id = ?"
            execute_query(query, (note_id,))
            
            return True
            
        except sqlite3.Error as e:
            raise Exception(f"Ошибка при удалении заметки с ID {note_id}: {e}")
    
    def count(self) -> int:
        """
        Возвращает общее количество заметок.
        
        Returns:
            Количество заметок в базе данных
        """
        try:
            query = "SELECT COUNT(*) as count FROM notes"
            rows = execute_query(query)
            return rows[0]['count'] if rows else 0
            
        except sqlite3.Error as e:
            raise Exception(f"Ошибка при подсчете заметок: {e}")